File: go/bitmap.go
package main

// Bitmap represents a rectangular grid of color-indexed pixels.
type Bitmap interface {
	Width() int
	Height() int
	Pixels() []int
}

// SimpleBitmap is a basic bitmap backed by a flat pixel array.
type SimpleBitmap struct {
	W, H   int
	pixels []int
}

func NewSimpleBitmap(w, h int, pixels []int) *SimpleBitmap {
	return &SimpleBitmap{W: w, H: h, pixels: pixels}
}

func (b *SimpleBitmap) Width() int    { return b.W }
func (b *SimpleBitmap) Height() int   { return b.H }
func (b *SimpleBitmap) Pixels() []int { return b.pixels }

File: zig/src/bitmap.zig
const std = @import("std");
const Color = @import("color.zig").Color;
const Allocator = std.mem.Allocator;

pub const Bitmap = struct {
    width: u32,
    height: u32,
    pixels: []const Color,

    pub fn getPixel(self: Bitmap, x: u32, y: u32) Color {
        return self.pixels[y * self.width + x];
    }
};

pub const OwnedBitmap = struct {
    bitmap: Bitmap,
    allocator: Allocator,

    pub fn deinit(self: *OwnedBitmap) void {
        self.allocator.free(self.bitmap.pixels);
    }
};

pub const Sprite = struct {
    bitmap: Bitmap,
    x: i32 = 0,
    y: i32 = 0,
    visible: bool = true,

    pub fn moveTo(self: *Sprite, x: i32, y: i32) void {
        self.x = x;
        self.y = y;
    }

    pub fn setVisible(self: *Sprite, visible: bool) void {
        self.visible = visible;
    }
};
